File: spiralMatrix.cpp
class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        vector<int> vec;

        if (matrix.empty())
            return vec;

        int top = 0, down = matrix.size() - 1, left = 0, right = matrix[0].size() - 1;
        int dir = 0;

        while (top <= down and left <= right)
        {
            if (dir == 0)
            {
                for (int i = left; i <= right; i++)
                {
                    vec.push_back(matrix[top][i]);
                }
                top++;
            }
            else if (dir == 1)
            {
                for (int i = top; i <= down; i++)
                {
                    vec.push_back(matrix[i][right]);
                }
                right--;
            }
            else if (dir == 2)
            {
                for (int i = right; i >= left; i--)
                {
                    vec.push_back(matrix[down][i]);
                }
                down--;
            }
            else
            {
                for (int i = down; i >= top; i--)
                {
                    vec.push_back(matrix[i][left]);
                }
                left++;
            }
            dir = (dir + 1) % 4;
        }

        return vec;
    }
};

File: spiralMatrix.py
class Solution:
    def spiralOrder(self, matrix):
        if not matrix:
            return []
        left, right, top, down, res = 0, len(matrix[0])-1, 0, len(matrix)-1, []
        while left <= right and top <= down:
            res.extend(matrix[top][left:right+1]) # left to right
            top += 1 
            for i in range(top, down+1): # top to down
                res.append(matrix[i][right])
            right -= 1
            if top <= down:
                res.extend(matrix[down][left:right+1][::-1]) # right to left
                down -= 1
            if left <= right:
                for i in range(down, top-1, -1): # bottom to up
                    res.append(matrix[i][left])
                left += 1
        return res